src/Swarm/Language/Syntax.hs

```diff
@@ -397,18 +396,21 @@ data Syntax = Syntax {sLoc :: Location, sTerm :: Term}
 data Location = Location {locStart :: Int, locEnd :: Int}
   deriving (Eq, Show, Data)
 
-emptyLoc :: Location
-emptyLoc = Location 0 0
+instance Semigroup Location where
+  Location s1 e1 <> Location s2 e2 = Location (min s1 s2) (max e1 e2)
+
+instance Monoid Location where
+  mempty = Location 0 0
```

Comment:
This is not a valid `Monoid`, since `mempty <> x /= x`, e.g. `Location 0 0 <> Location s e = Location (min 0 s) (max 0 e) = Location 0 e /= Location s e`.  `min` and `max` do not have identity elements, so I don't think there is any valid way to define `mempty`. 

I see we are using `mempty` in various places where we need some default `Location` which hasn't been specified yet.  Why don't we instead define `Location = NoLoc | Location Int Int` and then define `mempty = NoLoc`, and define the combining operation accordingly.  This is the standard way to make a `Semigroup` without an identity element into a `Monoid`, by adding a new special value to serve as `mempty`.  We could also do this by wrapping `Location` in `Maybe`, but it seems nicer to add our own custom constructor.

Reply:
Oops I see, I knew this was too easy. Thanks for the tip.